lib/puppet/functions/systemd/systemd_escape.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# @summary Escape strings by call the `systemd-escape` command in the background.
+Puppet::Functions.create_function(:'systemd::systemd_escape') do
+  # @param input Input string
+  # @param path Use path (-p) ornon-path  style escaping.
+  dispatch :escape do
+    param 'String', :input
+    optional_param 'Optional[Boolean]', :path
+    return_type 'String'
+  end
+
+  # rubocop:disable Style/OptionalBooleanParameter
+  def escape(input, path = false)
+    args = []
+
+    args.push('--path') if path
+
+    args.push(input)
+    exec_systemd(args)
+  end
+  # rubocop:enable Style/OptionalBooleanParameter
+
+  def exec_systemd(*args)
+    exec_args = { failonfail: true, combine: false }
+    escaped = Puppet::Util::Execution.execute(['systemd-escape', args], **exec_args)
```

Comment:
Relying on an external command should be avoided.  Can't we just improve the current `systemd::escape` function if it's current behavior is wrong?

Reply:
The current function is a different implementation. From my point of view, this is a breaking change, but I‘m fine with it.

Reply:
My personal opinion: Calling out to external tools should always be the last option. But Puppet already does that to systemd. Now it would be nice if there would be a systemd API binding for Ruby, but that doesn't really exist and that would require us to install a gem, which also sucks. We can keep the old function as well, so people can choose between using `systemd::systemd_escape` vs `systemd::escape`. I think that's fine.

Reply:
> so people can choose between

This was my first intension, too. And thats why I decide to use a different name. In combination with Defferred functions, systemd-escape is executed on the agent, not on the master.

I understand, calling an external tool is slow, but the puppet lange does not provide functions to translate characters to hex which is required to improve the existing function.

Reply:
The hex translation could be handled by another function/it could be rewritten as a Ruby function, but IMO that only makes it more complicated and there is no guarantee that our reimplementation works like the actual systemd-escape. That's why I think having both implementations is fine.